aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -0,0 +1,232 @@
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(ty_args.len(), 1);
+    debug_assert_eq!(args.len(), 1);
+    context.charge(AGGREGATOR_V2_READ_SNAPSHOT_BASE)?;
+
+    match ty_args[0] {
+        Type::U128 => {
+            let value = aggregator_snapshot_u128_info(&safely_pop_arg!(args, StructRef))?;
+            Ok(smallvec![Value::u128(value)])
+        },
+        Type::U64 => {
+            let value = aggregator_snapshot_u64_info(&safely_pop_arg!(args, StructRef))?;
+            Ok(smallvec![Value::u64(value)])
+        },
+        _ => {
+            // Check if the type is a string
+            if is_string_type(context, &ty_args[0])? {
+                let value = aggregator_snapshot_string_info(&safely_pop_arg!(args, StructRef))?;
+                let move_string_value = Value::struct_(Struct::pack(vec![Value::vector_u8(value)]));
+                return Ok(smallvec![move_string_value]);
+            }
+            // If not a string, return an error
+            Err(SafeNativeError::Abort {
+                abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
+            })
+        },
+    }
+}
+
+/***************************************************************************************************
+ * native fun native fun string_concat<Element>(before: String, snapshot: &AggregatorSnapshot<Element>, after: String): AggregatorSnapshot<String>;
+ **************************************************************************************************/
+
+fn native_string_concat(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(ty_args.len(), 1);
+    debug_assert_eq!(args.len(), 3);
+    context.charge(AGGREGATOR_V2_STRING_CONCAT_BASE)?;
+
+    let after = string_value_info(safely_pop_arg!(args, Struct))?;
+
+    let snapshot_value = match ty_args[0] {
+        Type::U128 => {
+            let value = aggregator_snapshot_u128_info(&safely_pop_arg!(args, StructRef))?;
+            Ok(value.to_string().into_bytes())
+        },
+        Type::U64 => {
+            let value = aggregator_snapshot_u64_info(&safely_pop_arg!(args, StructRef))?;
+            Ok(value.to_string().into_bytes())
+        },
+        _ => {
+            // Check if the type is a string
+            if is_string_type(context, &ty_args[0])? {
+                Ok(aggregator_snapshot_string_info(&safely_pop_arg!(
+                    args, StructRef
+                ))?)
+            } else {
+                Err(SafeNativeError::Abort {
+                    abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
+                })
+            }
+        },
+    }?;
+    let before = string_value_info(safely_pop_arg!(args, Struct))?;
+    let mut result = before.clone();
+    result.extend(&snapshot_value);
+    result.extend(&after);
```

Comment:
@georgemitenkov do you know, do we need to do any string checks/utf validations when appending like this (or use some move uility functions), or since all 3 should be guaranteed to be a valid string at this point, we can just concatenate them in rust?